---
title: "A tour of expressionAnalysis"
author: "Arianne Beauregard"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{A tour of expression Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(
	fig.align = "center",
	fig.height = 5.5,
	fig.width = 6,
	warning = FALSE,
	collapse = TRUE,
	dev.args = list(pointsize = 10),
	out.width = "90%",
	par = TRUE
)
knit_hooks$set(par = function(before, options, envir)
  { if(before && options$fig.show != "none") 
       par(family = "sans", mar=c(4.1,4.1,1.1,1.1), mgp=c(3,1,0), tcl=-0.5)
})
set.seed(1) # for exact reproducibility
```
       
## Introduction

`expressionAnalysis` is an R package aiming at identifying potential biomarkers
in a specific disease by analyizing gene expression data. It aims at making the
process of computing and visualising the data more accessible and efficient. 
**This document gives a tour of expressionAnalysis (version 0.1.0)**. 

To download **expressionAnalysis**, use the following commands:

```{r echo=T, results='hide', warning=FALSE, message=FALSE}
require("devtools")
devtools::install_github("ari-beau/expressionAnalysis", build_vignettes = TRUE)
library("expressionAnalysis")
```

To list all sample functions available in the package:
```{r echo=T, results='hide', warning=FALSE, message=FALSE}
ls("package:expressionAnalysis")
```


To list all sample datasets available in the package:
```{r echo=T, results='hide', warning=FALSE, message=FALSE}
ls("package:expressionAnalysis")
```


<br>

## Example usage of package

We will be using the datasets included in the package for this tutorial.
The datasets are from an ovarian cancer gene expression profiling experiment 
(Bowen N.J. et al., 2009). 
`OVExpression` contains the gene expression data, and `OVSample` includes 
sample information indicating if it is from a normal human ovary or from an 
ovarian cancer epithelial cell. Run `?OVExpression` and `?OVSample` for more
information
```{r}
head(OVExpression[ , 1:5])
head(OVSample)
```


### Normalization
We can use the function `exprNormalization` to normalize the data. In this
example, we will set the parameter *method* to "log" to use log2 transformation.
Run `?exprNormalization` for help and to read documentation for other methods. 
```{r}
logExpr <- exprNormalization(OVExpression, method = "log")
logExpr[1:5, 1:5]    # view first 5 rows and columns of normalized data
```


### Correlation
To visualize correlation between genes, we can use the `correlationPlot` 
function. Run `?correlationPlot` for help.
```{r}
correlationPlot(logExpr)
```



### Ranking differentially expressed genes
We can use the function `rankDEG` to rank the genes by differential expression 
between cases and controls. We will set the method parameter to "wilcoxon" to
use the Wilcoxon rank sum test. Run `?rankDEG` for help and  to read 
documentation for other methods.  
```{r}
rankWilcoxon <- rankDEG(expressionData = logExpr,
                        sampleData = OVSample,
                        case = "Ovarian cancer",                     
                        control = "Normal",
                        method = "wilcoxon")
head(rankWilcoxon)
```



To visualize differential gene expression, we use the function `exprPlot` to 
produce a boxplot. We are including the optional argument `genes` to select 
the genes we want to include in the plot. If not included, all genes will be 
included in the plot. Run `?exprPlot` for help.
```{r}
exprPlot(expressionData = logExpr, 
         sampleData = OVSample, 
         genes = c("PAX8", "DDR1", "THRA"))
```



## Package References

[Beauregard, A. (2022) expressionAnalysis. Unpublished.](https://github.com/ari-beau/expressionAnalysis)

<br>

## Other References

Bowen, N. J., Walker, L. D. E., Matyunina, L. V., Logani, S., Totten, K. A., Benigno, B. B., & McDonald, J. F. (2009). Gene expression profiling supports the hypothesis that human ovarian surface epithelia are multipotent and capable of serving as ovarian cancer initiating cells. BMC Medical Genomics, 2(1). https://doi.org/10.1186/1755-8794-2-71

Wei T, Simko V (2021). R package ‘corrplot’: Visualization of a Correlation Matrix. (Version 0.92), https://github.com/taiyun/corrplot.

Wickham H, François R, Henry L, Müller K (2022). dplyr: A Grammar of Data Manipulation. https://dplyr.tidyverse.org, https://github.com/tidyverse/dplyr.

Wickham H (2016). ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York. ISBN 978-3-319-24277-4, https://ggplot2.tidyverse.org.

Wickham H (2007). “Reshaping data with the reshape package.” Journal of Statistical Software, 21(12). https://www.jstatsoft.org/v21/i12/.

----

```{r}
sessionInfo()
```
